### PR TITLE
Introduce new method isClient($identifier) in JApplicationCms class

### PR DESCRIPTION
**Pull request by izharaazmi**
## Introduction

This PR introduces a new method `JApplicationCms::isClient($identifier)` as a substitute for `isSite` and `isAdmin` methods. Hence, remove usage of descendant class reference in the `JApplication` (legacy) and `JApplicationCms` for better OO structure and support more extensibility.

Since the current implementation essentially assumes its two subclasses `JApplicationSite` and `JApplicationAdministrator` which should be avoided as much possible.

This PR also enables more subclasses derived from `JApplicationCms` to make use of this method. Such as in the Installation app of the Joomla installer package, we currently **cannot** perform a test something like `$app->isInstallation();`. 
But this PR allows us to do the test like `$app->isClient('installation');`
## Feature Testing

Since this PR does not change any existing usage of existing methods the cms should not behave any different.

Run the following code in frontend, backend and installation applications respectively and match the output.

``` php
// Get instance of current application.
$app = JFactory::getApplication();

// Should give bool(true) in frontend, bool(false) otherwise.
var_dump($app->isClient('site'));
// var_dump($app->isClient(0)); 

// Should give bool(true) in backend, bool(false) otherwise.
var_dump($app->isClient('administrator'));
// var_dump($app->isClient(1));

// Should give bool(true) in Joomla installer, bool(false) otherwise.
var_dump($app->isClient('installation'));
// var_dump($app->isClient(2));
```

**EDIT**: Removed support for client id (integer) with 84163c5.

### CODE REVIEW

**Comment by mbabker** [comment]
1) I'd leave legacy JApplication alone, it's going away and already out of use in the CMS.
2) Personally, I would not accept mixed argument types on the new method.  I'd either use the client ID or the client name, the latter would be more easily recognized.  Methods that accept mixed parameter types (unless it's something acting as a data store like JObject or the Registry classes) looks like a code smell, you're essentially building two methods into one.

**Comment by izharaazmi** [comment]
@mbabker Well, then do you suggest to drop the integer argument support? Or split into two methods for each?
For a while I also thought to name that new method as just `is($identifier)` like `$app->is('site')` but not sure if it good. What do you suggest?

**Comment by mbabker** [comment]
I'd drop the integer and leave the name.  There are very few places that actually refer to the client IDs throughout the code base; the name is more favored.

**Comment by izharaazmi** [comment]
Exactly, I wonder why do we even support IDs, why not just deprecate without a substitute!

And, the PR... your suggestion sounds good to me. Should I make this changes or wait for more inputs from other fellow members?

File: libraries/cms/application/cms.php [hunk 1]
```diff
@@ -669,10 +669,11 @@ protected function initialiseApp($options = array())
 	 * @return  boolean  True if this application is administrator.
 	 *
 	 * @since   3.2
+	 * @deprecated  Use isClient('administrator') or isClient(1) instead.
 	 */
 	public function isAdmin()
 	{
-		return ($this->getClientId() === 1);
```
Comment (izharaazmi) [inline libraries/cms/application/cms.php]: @mbabker Currently the cms was using **only** `client_id`, therefore I used it and also added support for `client_name`. Should I now finally drop support for `client_id` match in my new proposed method?

Comment (Devportobello) [inline libraries/cms/application/cms.php]: Deprecated 4.0?

**Comment by izharaazmi** [comment]
Can we make this a part of 3.6 release? Somebody available to test this please?

**Comment by kevinscheithauer** [comment]
I have tested this item :white_check_mark: successfully on 84163c526d55141a3fc3b955d98e9e5f02f98981<br /><br />I have tested this issue @icampus PBF with the proposed change and it works as its supposed to. I followed the suggested steps and IsClient() returned the correct value in each case.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8971">issues.joomla.org/joomla-cms/8971</a>.</sub>

**Comment by brianteeman** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8971">issues.joomla.org/joomla-cms/8971</a>.</sub>

**Comment by rdeutz** [comment]
@izharaazmi could you have a look at the conflicts and resolve them, thanks!

**Comment by izharaazmi** [comment]
@rdeutz I have merged it to the 3.7.x branch resolving the conflicts. Please check.

**Pull request merged**